fms/utils/tensors.py

```diff
@@ -129,9 +132,253 @@ def cat(tensors, dim=0, *, out=None):
+        return _EXPANDABLE_HANDLED_FUNCTIONS[func](*args, **kwargs)
+
+
+_PAGED_HANDLED_FUNCTIONS = {}
+
+
+def _paged_implements(torch_function):
+    """Register a torch function override"""
+
+    def decorator(func):
+        functools.update_wrapper(func, torch_function)
+        _PAGED_HANDLED_FUNCTIONS[torch_function] = func
+        return func
+
+    return decorator
+
+
+R_LIST: TypeAlias = List[Union[int, "R_LIST"]]
+"""
+If we treat a tensor's underlying storage as a paged memory space,
+This method constructs the appropriate list of "page ids" to index
+into storage. The format is a nested list of page_ids where the
+outermost list is indexed on the first dynamic shape, etc.
+
+The second part of the tuple returned here is just an implementation
+detail used in recursion.
+"""
+
+
+def _create_page_ids(
+    dynamic_shapes: List[int], page_size: int, start_page: int
+) -> Tuple[R_LIST, int]:
+    if len(dynamic_shapes) == 1:
+        pages = int(math.ceil(dynamic_shapes[0] / page_size))
+        last_page = start_page + pages
+        return list(range(start_page, last_page)), last_page
+    else:
+        result: R_LIST = []
+        last_page = start_page
+        for _ in range(dynamic_shapes[0]):
+            to_add, last_page = _create_page_ids(
+                dynamic_shapes[1:], page_size, last_page
+            )
+            result.append(to_add)
+        return result, last_page
+
+
+class PagedStorage:
+    def __init__(
+        self,
+        static_shape,
+        dynamic_shape,
+        page_size=16,
+        initial_storage=None,
+        initial_page_ids=None,
+        dtype=None,
+        device=None,
+    ):
+        self.static_shape = static_shape
+        self.dynamic_shape = dynamic_shape
+        self.page_size = page_size
+        self.storage = initial_storage
+        if self.storage is None:
+            self.storage = torch.tensor(
+                [128 * page_size] + self.static_shape, device=device, dtype=dtype
+            )
+        elif initial_storage.numel() % (page_size * math.prod(self.static_shape)):
+            last_dynamic_len = dynamic_shape[-1]
+            # number of rows left on last dynamic dim to make a whole page
+            remaining_rows = page_size - (last_dynamic_len % page_size)
+            dyn_copy = dynamic_shape.copy()
+            dyn_copy[-1] = remaining_rows
+            buffer_shape = dyn_copy + static_shape
+            new_storage = torch.zeros(
+                buffer_shape, device=self.storage.device, dtype=self.storage.dtype
+            )
+            storage = torch.cat(
+                (initial_storage, new_storage), dim=len(dynamic_shape) - 1
+            )
+            self.storage = storage.view([-1, page_size] + self.static_shape)
+        else:
+            self.storage = initial_storage.view([-1, page_size] + self.static_shape)
+
+        pages = int(math.ceil(self.storage.numel() // page_size))
+
+        self.refcounts = [0] * pages
```

Comment:
do we want to keep refcount as a list or a tensor?

Reply:
I suspect this will be easier to manage on cpu, at least until we have some specific kernels. it should be easy to move to a tensor later as-needed too though